### ISSUE TITLE
Inconsistent Documentation about Optimizer.step(closure)

### ISSUE THREAD

**Issue by khdlr** [pytorch/pytorch#38006]
The docstring for the Optimizer class notes that when using the `.step` function with a closure, this closure should not change the parameter gradients:

https://github.com/pytorch/pytorch/blob/4e93844ab168ee0cf1aaa1b4712d6aad0e2972f8/torch/optim/optimizer.py#L174-L176

However, every example I've found seems do be doing exactly that, e.g. also on the `torch.optim` documentation on the same page:

```python
for input, target in dataset:
    def closure():
        optimizer.zero_grad()
        output = model(input)
        loss = loss_fn(output, target)
        loss.backward()
        return loss
    optimizer.step(closure)
```

Is the note about not changing the gradients in the closure outdated?

cc @vincentqb

**Comment by vincentqb** [comment]
Can you mention give an example that changes the closure where you would not expect it to?

Note that the documentation says "should not" which implies that it may still happen. :)

**Comment by cloud-oak** [comment]
It's more the other way around I guess.

After reading the documentation, I assumed that the optimizer (e.g. LBGFS) will do the `loss.backward` by itself. After all, my closure should not modify `.grad` unless otherwise specified. Needless to say, my model didn't learn anything at all in that experiment.

Do you understand why I think the remark about not changing the parameter's `.grad`s is misleading?

Is there actually a usecase where an optimizer is supposed to take a closure that doesn't change the parameter gradients?

**Comment by vincentqb** [comment]
Thanks for clarifying :) 

On one hand, the comment you referenced says not to modify the gradient in the closure.

https://github.com/pytorch/pytorch/blob/4e93844ab168ee0cf1aaa1b4712d6aad0e2972f8/torch/optim/optimizer.py#L174-L176

On the other, the example given does do the backward pass in the closure. And the [documentation](https://pytorch.org/docs/stable/optim.html) adds "Some optimization algorithms such as Conjugate Gradient and LBFGS need to reevaluate the function multiple times, so you have to pass in a closure that allows them to recompute your model. The closure should clear the gradients, compute the loss, and return it." I believe LBFGS is the only one that requires the `loss.backward()` to be used. Have you noticed other cases?

Thoughts?

**Comment by cloud-oak** [comment]
Even in "The closure should clear the gradients, compute the loss, and return it.", it isn't clear whether the closure should be calling `loss.backward()` (it should).

I made a short script that checks all optimizers in `torch.optim` (except for SparseAdam, because I don't know enough about sparse tensors): https://gist.github.com/cloud-oak/f1a14980f4cab66ed779cf7b73b44123

All optimizers change the target parameters if the closure calls `loss.backward()` (i.e. modifies the grads). However none of them changes the target parameters if the closure doesn't modify the grads.

**Comment by vincentqb** [comment]
> I made a short script that checks all optimizers in `torch.optim` (except for SparseAdam, because I don't know enough about sparse tensors): https://gist.github.com/cloud-oak/f1a14980f4cab66ed779cf7b73b44123

404 when clicking on the link :)

**Comment by khdlr** [comment]
That ought to teach me not to change my username... Sorry about that, updated the link in the above comment.

**Comment by nnop** [comment]
Any update on this? @vincentqb
